src/back/spirv/layout.rs

```diff
@@ -0,0 +1,216 @@
+            ext_inst_imports: vec![],
+            memory_model: vec![],
+            entry_points: vec![],
+            execution_modes: vec![],
+            debugs: vec![],
+            annotations: vec![],
+            type_declarations: vec![],
+            constants: vec![],
+            global_variables: vec![],
+            function_declarations: vec![],
+            function_definitions: vec![],
+        }
+    }
+
+    pub(crate) fn in_words(&self) -> Vec<u32> {
+        let mut words: Vec<Word> = vec![];
+
+        for capability in self.capabilities.iter() {
+            words.push(*capability);
+        }
+
+        for extension in self.extensions.iter() {
+            words.push(*extension);
+        }
+
+        for ext_inst_import in self.ext_inst_imports.iter() {
+            words.push(*ext_inst_import);
+        }
+
+        for memory_model in self.memory_model.iter() {
+            words.push(*memory_model);
+        }
+
+        for entry_point in self.entry_points.iter() {
+            words.push(*entry_point);
+        }
+
+        for execution_mode in self.execution_modes.iter() {
+            words.push(*execution_mode);
+        }
+
+        for debug in self.debugs.iter() {
+            words.push(*debug);
+        }
+
+        for annotation in self.annotations.iter() {
+            words.push(*annotation);
+        }
+
+        for type_declaration in self.type_declarations.iter() {
+            words.push(*type_declaration);
+        }
+
+        for constants in self.constants.iter() {
+            words.push(*constants);
+        }
+
+        for global_variables in self.global_variables.iter() {
+            words.push(*global_variables);
+        }
+
+        for function_declaration in self.function_declarations.iter() {
+            words.push(*function_declaration);
+        }
+
+        for function_definition in self.function_definitions.iter() {
+            words.push(*function_definition);
+        }
+
+        words
+    }
+}
+
+pub(crate) struct Instruction {
+    op: u32,
+    wc: u32,
+    type_id: Option<Word>,
+    result_id: Option<Word>,
+    operands: Vec<Word>,
+}
+
+impl Instruction {
+    pub(crate) fn new(op: Op) -> Self {
+        Instruction {
+            op: op as u32,
+            wc: 1, // Always start at 1 for the first word (OP + WC),
+            type_id: None,
+            result_id: None,
+            operands: vec![],
+        }
+    }
+
+    pub(crate) fn set_type(&mut self, id: Word) {
+        if self.type_id.is_none() {
+            self.wc += 1;
+        }
+        self.type_id = Some(id);
+    }
+
+    pub(crate) fn set_result(&mut self, id: Word) {
+        if self.result_id.is_none() {
+            self.wc += 1;
+        }
+        self.result_id = Some(id);
+    }
+
+    pub(crate) fn add_operand(&mut self, operand: Word) {
+        self.operands.push(operand);
+        self.wc += 1;
+    }
+
+    pub(crate) fn add_operands(&mut self, operands: Vec<Word>) {
```

Comment:
should be `&[Word]`

Reply:
Is there a reason to want an array and why would we want it to be reference? As `self.operands` is a `Vec<u32>` I cannot push a reference. I can dereference it with `*`, but I am not sure if that is a correct way.

Reply:
I assume the caller has a list of things to pass here. If the caller doesn't know the length statically, we could have an `Iterator<Item = Word>` here.
The problem with `Vec` in the argument is that it forces a heap allocation, while it doesn't care really. If it's `&[Word]` you can do `operands.iter().cloned()` to produce an iterator over `Word`, which is what you populate `self.operands` with.

Reply:
I'm not sure if I'm exactly following your thoughts. I changed the function now to:
```rust
pub(crate) fn add_operands(&mut self, operands: Vec<Word>) {
    self.operands.extend(operands.iter().cloned());
}
```
Would this suffice to circumvent the heap allocation, or does the argument needs to be refactored to either an `Iterator` or `&[ ]`?

**Side-note:** `add_operands` is only used for [`string_to_words`](https://github.com/Napokue/naga/blob/spirv-backend/src/back/spv/parser.rs#L97). Maybe there is some potential to refactor that one too and return something else than a `Vec`? I'm not sure yet though, as I've tried, but failed to do so up until now 😅

Reply:
It looks like this should be `fn add_operands(&mut self, operands: impl Iterator<Item = Word>)`, and `string_to_words` should return `impl Iterator<Item = Word>`